### ISSUE TITLE
shallow().dive() does not work with react-redux Provider

### ISSUE THREAD

**Issue by ThiefMaster** [enzymejs/enzyme#2202]
```jsx
it('should work', () => {
  const Component = () => {
    return <div>Hello World</div>;
  };
  const wrapper = shallow(
    <Provider store={mockStore()}>
      <Component />
    </Provider>
  ).dive();
  expect(wrapper).toMatchSnapshot();
});
```

This is just a simple example - please assume that `Component` needs to have access to the react-redux context so I can use `useSelector()` in there, so something like `.find(Component).dive()` won't work

### Current behavior
```js
exports[`<UserMenu /> hopefully works 1`] = `<Component />`;
```

### Expected behavior
```js
exports[`<UserMenu /> hopefully works 1`] = `
<div>
  Hello World
</div>
`;
```

#### API

- shallow

#### Version

| library             | version
| ------------------- | -------
| enzyme              | 3.10.0
| react               | 16.8.6
| react-dom           | 16.8.6
| react-test-renderer | -
| adapter (below)     | 1.14.0

#### Adapter

- enzyme-adapter-react-16

**Comment by chan-dra** [comment]
+1 here, doesn't seem to work for me either. 

```
const component = shallow(
      <Provider store={store}>
        <Test {...props} />
      </Provider>
    )
      .find(NavItem)
      .dive();
    expect(component).toMatchSnapshot();
```

throws this

```
Invariant Violation: could not find react-redux context value; please ensure the component is wrapped in a <Provider>
```

**Comment by JulianKl** [comment]
Same problem, are there any plans to fix this? Also, does anyone have a workaround that can be used until it is fixed?

**Comment by markkra** [comment]
Big corporate app - same issue. If we want to move forward and start using Redux with Hooks we need this fixed. A work around would be great. Right now we only have one component that uses .dive() heavily and if someone figures out a workaround that would be great.

**Comment by MikeSpitz** [comment]
Any update on how best to get around this? Or if this will be tackled?

**Comment by michielmetcake** [comment]
Same here 💃

**Comment by ljharb** [comment]
Try with the `wrappingComponent` option?

**Comment by idangozlan** [comment]
Same here. seems like enzyme and redux hooks are not getting along.

**Comment by KenLui** [comment]
@idangozlan, shallow() doesn't work with React's useContext hook. If you're using a version of react-redux higher than 5.1.1, you're going to have a bad time dealing with contexts. See [this comment](https://github.com/airbnb/enzyme/issues/2176#issuecomment-532361526) for more some workarounds.

```js
// Bare component.
const Component = () => {
  return <div>Hello World</div>;
};

// Redux connected component.
const ConnectedComponent = connect()(Component);

it('inject store work', () => {
  // Inject store directly into connected component.
  const wrapper = shallow(
    <ConnectedComponent store={mockStore()} />
  );
  expect(wrapper).toMatchSnapshot();
});
```

**Comment by idangozlan** [comment]
@KenLui it's still not helpful, as the component not getting rendered... it render the snapshot as:
```
<Test store={{...}} dispatch={[Function: dispatch]} />
```

**Comment by ljharb** [comment]
enzyme doesn't have snapshot support, nor recommend snapshot testing - the closest thing is `.debug()` (or `.debug({ verbose: true })`). If you're having trouble with snapshots, you'll need to file an issue on whatever is generating them.

**Comment by idangozlan** [comment]
right, but even when running `.debug()` it's rendering the result above, shouldn't it render the `Test` component with its markup?

**Comment by ljharb** [comment]
@idangozlan with shallow, you need one `.dive` per HOC - so since you have `connect()`, you need `wrapper.dive().debug()` to see the contents of your inner component.

**Comment by idangozlan** [comment]
@ljharb make sense, but then @KenLui is worthless. getting `Invariant Violation: could not find react-redux context value; please ensure the component is wrapped in a <Provider>`. :)

Hopefully, we will have a solution to this issue soon.

**Comment by ljharb** [comment]
@idangozlan that's why i suggested using the `wrappingComponent` option:
```js
const wrapper = shallow(<ComponentThatNeedsContext>, {
  wrappingComponent: ContextProvider
});
wrapper.debug()
```

**Comment by idangozlan** [comment]
Still happening (`Invariant Violation: could not find react-redux context value; please ensure the component is wrapped in a <Provider>`).

The sample code is:

Test.js:
```
import React from 'react';
import { useSelector } from 'react-redux';

export default function Test() {
  const user = useSelector(state => state.auth.user);

  console.log(user);
  // const user = 'test user';
  // console.log(user);

  return (
    <div className="D">
      <div className="C">
        <div className="B">
          <div className="A">
My Test
            {user || ''}
          </div>
        </div>
      </div>
    </div>
  );
}
```

The test:
```
import React from 'react';
import PropTypes from 'prop-types';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import Test from './Test';

const mockStore = configureStore([]);
const defaultStore = mockStore({
  auth: { user: 'Idan' },
});

function MyProvider(props) {
  const { children } = props;

  return (
    <Provider store={props.store}>
      {children}
    </Provider>
  );
}

MyProvider.propTypes = {
  children: PropTypes.node.isRequired,
  store: PropTypes.object,
};
MyProvider.defaultProps = {
  store: defaultStore,
};

describe('My Desc', () => {
  it('My Test', () => {
    expect.assertions(1);

    const wrapper = shallow(<Test />, {
      wrappingComponent: MyProvider,
    });
    // const provider = wrapper.getWrappingComponent();
    // provider.setProps({ store: defaultStore });

    console.log(wrapper.debug());
    expect(1).toEqual(1);
  });
});
```

**Comment by ThiefMaster** [comment]
Indeed, I tried it myself again and even without snapshots etc, the problem is that `wrappingComponent` doesn't seem to work with `useContext` (which react-redux uses internally)...

**Comment by ljharb** [comment]
Yes, that's very true. React hooks don't provide any hooks for testing purposes, so there's nothing for enzyme to interact with. You'd have to use the non-hook version of react-redux to be able to test it properly.

**Comment by idangozlan** [comment]
@ljharb do you expect any upcoming support for case like that in enzyme?

**Comment by ljharb** [comment]
@idangozlan i certainly hope so - but it would require react to expose API hooks into how hooks work, and/or to implement support for such things in the shallow renderer.

**Comment by sunilb0575** [comment]
Any solution for this issue, out there?

**Comment by skyboyer** [comment]
if enzyme did not filter context by `.contextType`(check https://github.com/airbnb/enzyme/blob/master/packages/enzyme-adapter-utils/src/Utils.js#L321) I believe we could pass store directly as options: `.dive({context: { store }})`. By now any context data passed explicitly is vanished by `getMaskedContext()`.

**Comment by sunnyofhell** [comment]
Same Problem. Found a temporary solution though

https://stackoverflow.com/questions/59191129/enzyme-jest-react-testing-shallow-connected-component-with-react-redux-6

**Comment by sunnyofhell** [comment]
> Any solution for this issue, out there?

https://stackoverflow.com/questions/59191129/enzyme-jest-react-testing-shallow-connected-component-with-react-redux-6

**Comment by jonluca** [comment]
Bump on this issue - related to #2174 

Currently looking like theres no way to shallow render when using the useContext hook with enzyme. Has there been any progress on this?

**Comment by vojtatranta** [comment]
Hey,
just came across same issue. I concluded that for test purposes injecting `store` directly to the connected component is OK. The only problem is Typescript typing because types do not allow passing store directly as a prop. Otherwise, it is OK.

**Comment by chris-fran** [comment]
@vojtatranta how do inject it when the connected component is actually a child?

```jsx
const wrapper = shallow(<MyComp />);

connst connected = wrapper.find('Connect(Something)');

connected.dive()  // doesnt work
connected.dive({ context: { store } })  // doesnt work
connected.shallow()  // doesnt work
connected.shallow({ context: { store } })  // doesnt work
```

**Comment by vojtatranta** [comment]
@chris-fran pass the store as prop to connect HOC.

**Comment by stella1013** [comment]
I was having issues with TypeScript and React-Redux Provider. I don't know if this helps anyone but I was able to get a DOM snapshot via the below code.

library | version
-- | --
enzyme | 3.11.0
react | 16.12.0
react-redux | 7.1.7
react-dom | 16.12.0
react-test-renderer | 16.9.2
enzyme-adapter-react-16| 1.15.2
redux-mock-store | 1.5.4
typescript | 3.7.2

testUtils.ts
```ts
export const storeFactory = (initialState: any) => {
	return createStore(rootReducer, initialState);
};
```

test.tsx
```ts
const connectedSetup = (initialState={}) => {
    const store = storeFactory(initialState);
    const wrapper = shallow(<ConnectedSplashScreen store={store}/>).dive().dive();
	console.log(wrapper.debug({verbose:true}));
	return wrapper;
};
```

Though I do receive a Type error on 'store' on this line `<ConnectedSplashScreen store={store}/>`
```
(JSX attribute) store: Store<CombinedState<{
    clientState: ClientState;
    myForm: FormState;
    data: DataState;
}>, setVoucherUrlAction | ... 22 more ... | dataSetAction>
Type '{ store: Store<CombinedState<{ clientState: ClientState; myForm: FormState; data: DataState; }>, setVoucherUrlAction | ... 22 more ... | dataSetAction>; }' is not assignable to type 'IntrinsicAttributes & Pick<ReduxType, never>'.
  Property 'store' does not exist on type 'IntrinsicAttributes & Pick<ReduxType, never>'.ts(2322)
```

**Comment by Quirksmode** [comment]
Same issue for me, such a shame as replacing connect with the hooks removes so much boilerplate

**Comment by stroiman** [comment]
I tried mocking out react-redux's `useStore` hook,

```
import * as reduxHooks from 'react-redux/lib/hooks/useStore';

sinon.stub(reduxHooks, 'useStore').returns(myMockedStore)
```

But my component under test didn't pick up the stubbed version.

Maybe using proxyquire can solve the mocking issue as described here: https://sinonjs.org/how-to/link-seams-commonjs/

But that was more work than I wanted to go to for something I assume will get fixed eventually? I will stick with `mount` for now.

**Comment by Ba-stian** [comment]
I have the similiar issue: when I'm trying to access into connected component, then find API return 0 nodes. 

```
import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import ReportButton from './ReportButton';

configure({ adapter: new Adapter() });
const mockStore = configureStore();
const store = mockStore();
let container;


describe('Report Button', () => {
	beforeEach(() => {
		const mockProps = {
			fullReport: false,
			period_tariff_is_active: true,
			dataId: '',
			loading: false,
			isEmpty: false,
			count: 100,
			filter: {},
			category: [],
			type: '',
		};
		container = shallow(<Provider store={store}><ReportButton {...mockProps}/></Provider>).dive();
	});

	afterEach(() => {
		container = null;
	});

	it('should render component without crashing', () => {
		expect(container).toMatchSnapshot();
	});
	it('should find report button', () => {
		expect(container.find('.button').length).toEqual(1);
	});
});
```

**Comment by Ba-stian** [comment]
So, I want share my observations.

I noticed that the described problem hapend, when we use mapDispatchToProps/mapStateToProps + connect in functional component, then snapshot test return only ContextProvider wrapper and nothing inside him.

So, if you replace mapDispatchToProps/mapStateToProps with react-redux hooks useDispatch/useSelector, at first, you don't need to wrap your Component in connect HOC, and snapshot will return ContextProvider(Component).

Also, when you write Component in class-style and use mapDispatchToProps/mapStateToProps + connect, then snapshot return ContextProvider(Component) too.

**Comment by reyapr** [comment]
use `mount` instead of `shallow`

**Comment by ljharb** [comment]
@Ba-stian don't use mount; use the `wrappingComponent` option to pass Provider, and pass `<ReportButton />` directly to `shallow`.

**Comment by Ba-stian** [comment]
@ljharb I rewrote my test like that

```
import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ReportButton from "./ReportButton";
import {Provider} from "react-redux";
import {store} from "../../store/createdStore";
import configureStore from 'redux-mock-store';


configure({ adapter: new Adapter() });
let container;
const ShallowProvider = props => {
    const {children, customStore} = props;
    return (
        <Provider store={customStore || store}>
            {children}
        </Provider>
    )
};
const mockStore = configureStore();

describe('Report Button', () => {
    beforeEach(() => {
        const mockProps = {
            fullReport: false,
            period_tariff_is_active: true,
            dataId: '',
            loading: false,
            isEmpty: false,
            count: 100,
            filter: {},
            category: [],
            type: '',
        };
        container = shallow(<ReportButton {...mockProps}/>, {
            wrappingComponent: ShallowProvider
        });
        const provider = container.getWrappingComponent();
        provider.setProps({ customStore: mockStore})
    });

    afterEach(() => {
        container = null;
    });

    it('should render component without crashing', () => {
        expect(container).toMatchSnapshot();
    });
});

```

And now I get undefined in my snapshot((

**Comment by Leessonomy** [comment]
This is very actual problem. I spent several hours to solve it and i just wrote mount() instead of shallow()

**Comment by Ba-stian** [comment]
For me using mount return undefined in snapshot too

**Comment by ljharb** [comment]
`provider.setProps()` won't likely work without a `container.update()` after it.

**Comment by ljharb** [comment]
Separately, enzyme doesn't support or recommend using snapshot testing; the closest is `.debug()`.

**Comment by Ba-stian** [comment]
I add `container.update()` below `provider.setProps()`
and I still get undefined

```
    TypeError: Cannot read property 'debug' of undefined

      46 | 
      47 |     it('should render component without crashing', () => {
    > 48 |         console.log(container.debug())
         |                               ^
      49 |     });
      50 |     it('should find button in component', () => {
      51 |         expect(container.find('.button').length).toEqual(1);

```

in second test

```
it('should find button in component', () => {
        expect(container.find('.button').length).toEqual(1);
    })
```

I get this

```
    could not find react-redux context value; please ensure the component is wrapped in a <Provider>

       9 | const ReportButton = ({fullReport, dataId, count, type}) => {
      10 | 
    > 11 |     const period_tariff_is_active = useSelector(state => state.auth.user && state.auth.user.period_tariff_is_active);
         |                                     ^
      12 |     const lang = useSelector(state => state.auth.user && state.auth.user.lang);
      13 |     const filter = useSelector(state => state.selectFilter);
      14 |     const loading = useSelector(state => state.loading.loading);

```

**Comment by renjithspace** [comment]
> @idangozlan that's why i suggested using the `wrappingComponent` option:
> 
> ```js
> const wrapper = shallow(<ComponentThatNeedsContext>, {
>   wrappingComponent: ContextProvider
> });
> wrapper.debug()
> ```

Using `wrappingComponent` is got worked 👍

**Comment by anosov-otg** [comment]
@renjithspace could you please explain that code snippet. What is ContextProvider in this case?

**Comment by renjithspace** [comment]
@anosov-otg React redux provider

**Comment by ljharb** [comment]
Closing; if `wrappingComponent` doesn't work for anyone, please file a new issue.

**Comment by anosov-otg** [comment]
`wrappingComponent` doesn't work for me, but these options do: https://github.com/enzymejs/enzyme/issues/1002#issuecomment-421535999

**Comment by ljharb** [comment]
@anosov-otg the first and third are fine, I'd expect the second to work with `wrappingComponent` as the Provider instead.

**Comment by dschinkel** [comment]
>const wrapper = shallow(<ComponentThatNeedsContext>, {
  wrappingComponent: ContextProvider
});

what about provider with store

**Comment by ljharb** [comment]
@dschinkel you'd use the `wrappingComponentProps` option to supply props to the `Provider`.

**Comment by dschinkel** [comment]
yea this is kinda nasty TBH https://enzymejs.github.io/enzyme/docs/api/ShallowWrapper/getWrappingComponent.html

thanks though

**Comment by dschinkel** [comment]
tried it, I get `Cannot read property 'getState' of undefined`
```js
it('contains expected actions', () => {
  const homePage = shallow(<HomePageContainer />, {
		wrappingComponent: Provider,
		wrappingComponentProps: store
	});

  expect(homePage.props().fetchFeaturedCompanies).to.exist;
  expect(homePage.props().fetchCompanies).to.exist;
});
```
I won't derail this and can open another issue if need be.

**Comment by ljharb** [comment]
`wrappingComponentProps` needs to be a props object. try `{ store }`.

**Comment by dschinkel** [comment]
referring to this for further discussion (2449)[https://github.com/enzymejs/enzyme/issues/2449]

**Comment by dschinkel** [comment]
https://github.com/enzymejs/enzyme/issues/2448#issuecomment-697105718

**Comment by renjithspace** [comment]
@dschinkel @anosov-otg We can tackle this in different ways.

### Solution 1:
Create an HOC for your provider, here is an example with redux:

```jsx
// StoreProvider.js

import { Provider } from 'react-redux'
import store from 'path/to/store'

function StoreProvider() {
  return (
    <Provider store={store}>
      {this.props.children}
    </Provider>
  )
}
```

And use that in tests with `wrappingComponent`:
```jsx
// ComponentThatNeedsProvider.test.js

import { shallow } from 'enzyme'
import StoreProvider from 'path/to/StoreProvider'
import ComponentThatNeedsProvider from 'path/to/ComponentThatNeedsProvider'

const wrapper = shallow(<ComponentThatNeedsProvider>, {
  wrappingComponent: StoreProvider
});
```

### Solution 2:
You can directly pass provider props with `wrappingComponentProps` like below without creating an HOC for provider as @ljharb mentioned.

```jsx
// ComponentThatNeedsProvider.test.js

import { shallow } from 'enzyme'
import { Provider } from 'react-redux'
import store from 'path/to/store'
import ComponentThatNeedsProvider from 'path/to/ComponentThatNeedsProvider'

const wrapper = shallow(<ComponentThatNeedsProvider>, {
  wrappingComponent: Provider,
  wrappingComponentProps: { store }
});
```

If you prefer [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) and if you have multiple providers with a bunch of props to pass, go with **Solution 1** 👍

**Comment by dschinkel** [comment]
@stella1013
here's an answer to that error you got about store and how to resolve it
https://medium.com/@DaveSchinkel/setting-up-mocha-js-to-run-typescript-42f3099450eb

**Comment by dschinkel** [comment]
@renjithspace

yea I have tried that but no luck with `wrappingComponentProps: { store }`

**Comment by ajGingrich** [comment]
The wrapping component only works for me when I use a full mount. 

```
const wrapper = mount(<MyComponent />, {
    wrappingComponent: Provider,
    wrappingComponentProps: { store },
})
```

works but when using `shallow` it doesn't. That defeats the whole purpose, I could just do a full mount on the provider in the first place which is want I want to avoid.

**Comment by ljharb** [comment]
@ajGingrich i agree; the goal is that it works with `shallow`. If you're still having that trouble, it'd be great if you filed a new issue about it so we could try to work through it.

**Comment by marquezmanu** [comment]
I will leave my experience here it might help any folk with a similar project setup.

In the project I work, we are using enzyme v`3.11.0`, react-redux v`7.2.4` and typescript v`4.2.4`. The problem we had was with the tests for a connected component,  when wrapping the component with the react-redux Provider, the shallow output was not the expected one, and when trying to dive into it, again the results was not the expected ones, similar of what it is described for several folks above. 
The solution for us was to pass directly to the Connected component the fake store, and not wrap it with the react-redux Provider or any custom provider, with that typescript was yelling at us since store was not a valid prop, so to fix that we included in the typescript config this little snippet of code.

- In the file `tsconfig.json` include the following snippet. 

```ts
declare namespace JSX {
  interface IntrinsicAttributes {
    store?: unknown;
  }
}
```

With that now typescript is happy and we as well.

And below a small example of how we do the shallow test for the connected components.

```tsx
import configureMockStore from 'redux-mock-store';
const mockStore = configureMockStore();
const storeFake = mockStore({ foo: 'bar' });

it('should render the wrapped component with the connected properties', () => {
  const wrapper = shallow(<Component store={storeFake} />);
  expect(toJson(wrapper)).toMatchInlineSnapshot(`
  <ContextProvider
    value={null}
  >
    <Component
      dispatch={[Function]}
      foo="bar"
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    />
  </ContextProvider>
  `);
}
```

Note: Not a native english speaker sorry for any typo or for a bad expressed idea. Please quote me if it is needed a bit of more  explanation. 

Please check this [post](https://daveschinkel.medium.com/setting-up-mocha-js-to-run-typescript-42f3099450eb) to have a more detailed explanaition on how to include the ts type extension in your project.

**Comment by coskunuyar** [comment]
I had the same problem using dive twice solved my issues.Thats interesting.

`const setup = (initialState = {}) => {
  const store = storeFactory(initialState);
  return shallow(<Input store={store} />)
    .dive()
    .dive();
};`

**Comment by ljharb** [comment]
@coskunuyar one dive per HOC, yes.

**Comment by joan-saum** [comment]
How come this issue has still no solution? React hooks has been introduced in 2019, and it is still not possible to access the redux store with react hooks during a shallow render..

**Comment by ljharb** [comment]
@joan-saum and it will still have no solution in 2119, since it’s blocked on the react team providing hooks for testing/reaching into hooks.

**Comment by Pragalbhmandaokar** [comment]
_Issue mentioning_
**This seems to me the right way to write provider but the output should be, rendered component but I am only getting <ComponentName>**
```
const store = configureReducer();
let wrapper = ShallowWrapper
describe('landing page UI check', () => {
  beforeEach(()=>{
    wrapper = shallow(
      <Provider store={store}>
        <ComponentName />
      </Provider>
    ).dive();
  })
  it('format page', async () => {
    expect(wrapper).toMatchSnapshot();
  });
});
```
